In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
import warnings

warnings.filterwarnings("ignore")


In [ ]:
from sklearn.model_selection import train_test_split,GridSearchCV,RandomizedSearchCV, KFold

# Estimators
from sklearn.linear_model import LogisticRegression,Perceptron,SGDClassifier
from sklearn.svm import SVC

from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import BaggingClassifier,RandomForestClassifier,GradientBoostingClassifier,AdaBoostClassifier

from sklearn.metrics import confusion_matrix
from sklearn.model_selection import GridSearchCV

# Metrics
from sklearn.metrics import accuracy_score,recall_score,f1_score,precision_score
from sklearn.metrics import ConfusionMatrixDisplay, classification_report


#Preamble: Load the dataset and examine it.

### Notes:

* Maximum marks : 80

* This exam consists of a **CLASSIFICATION** problem.  

* The target is the '**Heart_Disease**' column.

* Random state should be taken as 64 wherever applicable.

* For NAT type of question if nothing is mentioned, Enter the answer accurately upto 3 decimal places.

* The dataset is already preprocessed, i.e. no missing values and numerical and categorical features are scaled/encoded accordingly.

* Sklearn assumes class 1 as positive class. Therefore, precision, recall are calculated for class 1.

###Q1. [marks : 0] Which dataset are you using for this exam?

Options:

A) V1

B) V2

C) V3

D) V4

E) V5


Answer:
V1: A, V2:B, V3: C, V4:D, V5:E

In [ ]:
df = pd.read_csv('/content/V2.csv')

In [ ]:
df.shape

(12499, 19)

In [ ]:
Data = df.copy()

In [ ]:
Data.head()
Data.columns

Index(['General_Health', 'Checkup', 'Exercise', 'Skin_Cancer', 'Other_Cancer',
       'Depression', 'Diabetes', 'Arthritis', 'Sex', 'Age_Category',
       'Height_(cm)', 'Weight_(kg)', 'BMI', 'Smoking_History',
       'Alcohol_Consumption', 'Fruit_Consumption',
       'Green_Vegetables_Consumption', 'FriedPotato_Consumption',
       'Heart_Disease'],
      dtype='object')

##Q2 [Marks =3] Break the dataset into features(`X`) and label (`y`), where the column `Heart_Disease` goes to `y` and the rest of the columns go to `X`. What proportion of data points belongs to label 1?  [NAT]

Answer:

V1: 0.58 Range: [0.54, 0.62]

V2: 0.58 Range: [0.54, 0.62]

V3: 0.58 Range: [0.54, 0.62]

V4: 0.58 Range: [0.54, 0.62]

V5: 0.58 Range: [0.54, 0.62]


In [ ]:
X = Data.drop(['Heart_Disease'], axis = 1)
y = Data['Heart_Disease']

In [ ]:
Data.Heart_Disease.value_counts()/(len(Data))

1    0.587407
0    0.412593
Name: Heart_Disease, dtype: float64

##Q3 [Marks : 3] Split the dataset into train and test dataset into the `80:20` ratio while keeping `random_state =64`. How many examples are there in training dataset?


Ans: $9999$ (No range is required) for all version

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 64)

# print(X_train.shape)
#print(X_test.shape)
print(y_train.shape)
#print(y_test.shape)

(9999,)


##Q4 [4 marks][NAT] Take `LogisticRegression` estimator with following parameters for training:
  - Use `lbfgs` as solver
  - Set random state to be equal to 64
  - Tolerance for stopping criteria to be 1e-4
  - inverse of regularization parameter, `C` = 0.1
  - Maximum number of iterations taken for the solvers to converge to be 100

Enter the f1 score for the given model using test set(X_test, y_test)


V1: 0.88 (Range: 0.82 , 0.92)

V2: 0.86 (Range: 0.8 , 0.9)

V3: 0.88 (Range: 0.82 , 0.92)
  
V4: 0.88 (Range: 0.82 , 0.92)
   
V5: 0.88 (Range: 0.82 , 0.92)


In [ ]:
logistic = LogisticRegression(solver='lbfgs',
                              tol=1e-4,
                              max_iter=100, C = 0.1,
                              random_state =64)
logistic.fit(X_train, y_train)

y_pred = logistic.predict(X_test)

f1_score(y_test,y_pred)

0.8649186856953204

## Q 5 [4 marks] [NAT]: What is the intercept (bias term in the decision function) learnt by the the above model?

**Answer**

V1: $-0.09$ Range: [-0.12, -0.06]

V2: $-0.08$ Range: [-0.11, -0.05]

V3: $-0.11$ Range: [-0.15, -0.07]

V4: $-0.09$ Range: [-0.12, -0.06]

V5: $-0.10$ Range: [-0.14, -0.06]

In [ ]:
logistic.intercept_

array([-0.08906489])

# **(Common Instructions for Q6 and Q7)**

Take LogisticRegression(random_state = 64) estimator with GridSearchCV. Hyperparameter tuning to be done over the following parameters:

- solver as 'lbfgs' or 'sag'

- Maximum number of iterations taken for the solvers to converge to be [100, 200, 500]

- value of inverse reularization parameter `C` to be [0.01, 0.1, 1, 10]

- Use cross validation = 3


Use the best model from above hyper parameter turing process to answer following questions:

In [ ]:
LR_clf = LogisticRegression(random_state = 64)

LR_params = {'solver':['lbfgs','sag'],
          'C':[0.01, 0.1, 1, 10],
          'max_iter':[100, 200,500]}

LR_gscv = GridSearchCV(LR_clf,
                    LR_params,
                    cv= 4,)

LR_gscv.fit(X_train, y_train)

GridSearchCV(cv=4, estimator=LogisticRegression(random_state=64),
             param_grid={'C': [0.01, 0.1, 1, 10], 'max_iter': [100, 200, 500],
                         'solver': ['lbfgs', 'sag']})

## Q6:[Marks: 4] [NAT] Enter the value of `C` of the best estimator you got after training with GridSearchCV.



Answer:

 V1: $1$

 V2: $10$

V3: $10$

V4: $1$

V5: $10$



In [ ]:
LR_gscv.best_params_

{'C': 10, 'max_iter': 200, 'solver': 'lbfgs'}

## Q7. [Marks: 4][NAT] Enter the value of the precision on the test set using the best model:

 Answer:

V1: 0.86 (Range : 0.81 ,0.91)

V2: 0.854 (Range: 0.8, 0.9)

V3: 0.87 (Range: 0.82, 0.92)
  
V4: 0.86 (Range: 0.81, 0.91)  
   
V5: 0.86 (Range: 0.81, 0.91)

In [ ]:
y_pred = LR_gscv.best_estimator_.predict(X_test)
precision_score(y_test,y_pred)

0.854521625163827

##Q8  [6 marks] [NAT] Use  `SGDClassifier` on the training dataset (`X_train` and `y_train`) to train the model. Use the following parameters:

1. `log_loss`is the loss function to be used
2. apply ridge regularization,
3. maximum number of iterations is 10
4. constant learning rate of 0.01,
5. regularization rate value is 0.001,
6. Take random_state=64.
7. Set warm_start as False

Note : Please ignore the convergence warning.

Using above model, calculate and write the correct value of f1_score for the test set.

Ans:

V1. 0.78 (Range: 0.74, 0.82)

V2: 0.855 (Range: 0.8, 0.9)

V3: 0.858 (Range: 0.8, 0.9)

V4: 0.862 (Range: 0.81, 0.91)

V5: 0.826 (Range: 0.78, 0.86)



In [ ]:
logistic_sgd = SGDClassifier(loss='log_loss',
                        penalty='l2',
                        max_iter=10,
                        warm_start=False,
                        eta0=0.01,
                        alpha=0.001,
                        learning_rate='constant',
                        random_state=64)

logistic_sgd.fit(X_train,y_train)

SGDClassifier(alpha=0.001, eta0=0.01, learning_rate='constant', loss='log_loss',
              max_iter=10, random_state=64)

In [ ]:
f1_score(y_test,  logistic_sgd.predict(X_test))

0.8557487038731321



## Q9. [4 marks] [MCQ] Use Gridsearchcv with KNeighborsClassifier estimator, `f1` as scoring parameter, cv= 5. Consider [5,7,9,11] as K values to be examined.
* Consider following parameters for KNeighborsClassifier:

 - Use Euclidean distance metric


- Keep other parameter values as default values.

What is the best value of K you obtained using the above instructions?



Ans:

V1: $9$

V2: $11$

V3: $7$

V4: $11$

V5: $7$

In [ ]:
knn_clf = KNeighborsClassifier(metric = 'euclidean')

knn_gscv = GridSearchCV(knn_clf,
                  param_grid= {'n_neighbors':[5, 7 , 9, 11]},
                  scoring='f1',
                  cv=5)

knn_gscv.fit(X_train,y_train)
knn_gscv.best_params_

{'n_neighbors': 11}

# **(Common Instructions for Q10 and Q11)**

Take DecisionTreeClassifier(random_state = 64) estimator with GridSearchCV. Hyperparameter tuning to be done over the following parameters:

- Criterion as 'entropy' or 'gini'

- Splitter as 'random' or 'best'

- Minimum number of samples per leaf as [6,8,10, 12, 14]

- Maximum depth as [5,6, 7, 8]

- Minimum mumber of samples per node to split as [16, 18, 20, 22]

- Use cross validation = 3


Use the best model from above hyper parameter turing process to answer following questions:

In [ ]:

dt_clf = DecisionTreeClassifier(random_state = 64)

params = {'criterion':['gini','entropy'],
          'splitter':['best','random'],
          'min_samples_leaf': [6,8,10, 12, 14],
          'max_depth':[5,6, 7, 8], 'min_samples_split': [16, 18, 20, 22]}

# t=KFold(n_splits = 4, random_state = 64)
dt_gscv = GridSearchCV(dt_clf,
                    params,
                    cv= 3,)

dt_gscv.fit(X_train, y_train)

GridSearchCV(cv=3, estimator=DecisionTreeClassifier(random_state=64),
             param_grid={'criterion': ['gini', 'entropy'],
                         'max_depth': [5, 6, 7, 8],
                         'min_samples_leaf': [6, 8, 10, 12, 14],
                         'min_samples_split': [16, 18, 20, 22],
                         'splitter': ['best', 'random']})

## Q10 [5 marks] What is the depth of the best estimator?

**Answer**

V1: $7$

V2: $7$

V3: $7$

V4: $7$

V5: $8$


In [ ]:
dt_gscv.best_estimator_.get_depth()

7

## Q11 [5 marks] Find the number of leaves in the best estimator.

**Answer**

V1: 99

V2: 101

V3: 95

V4: 103

V5: 148

In [ ]:
dt_gscv.best_estimator_.tree_.n_leaves

101

# **(Common Instructions for Q12, Q13 and Q14)**

Suppose we train a decision tree classifier using the following two approaches:

Approach 1. Do not grow the full size tree by setting the parameters as:

  - max depth of the tree to be 10.
  - minimum samples in a node to split to be 50.

Approach 2. Use cost complexity pruning by setting the cost complexity parameter as 0.01.

NOTE:
- Use random state to be 64 in both approaches.
- Use `entropy` criterion in both approaches.


In [ ]:
# DT_clf = DecisionTreeClassifier(criterion = 'entropy',  random_state = 64)
# # DT_clf.fit(X_train, y_train)
# # DT_clf.get_depth()

# params = {'ccp_alpha': [0.01, 0.0001,  0.1, 1, 2, 3, 10, 15]}

# # t=KFold(n_splits = 4, random_state = 64)
# gscv = GridSearchCV(DT_clf,
#                     params,
#                     cv= 4,)

# gscv.fit(X_train, y_train)
# gscv.best_params_

In [ ]:
DT_clf_1 = DecisionTreeClassifier(criterion = 'entropy', max_depth = 10, min_samples_split = 50, random_state = 64)
DT_clf_1.fit(X_train, y_train)

DecisionTreeClassifier(criterion='entropy', max_depth=10, min_samples_split=50,
                       random_state=64)

In [ ]:
DT_clf_2 = DecisionTreeClassifier(criterion = 'entropy', ccp_alpha = 0.01, random_state = 64)
DT_clf_2.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.01, criterion='entropy', random_state=64)

In [ ]:
print(DT_clf_1.score(X_test, y_test))
print(DT_clf_2.score(X_test, y_test))

0.8364
0.8052


##Q12 [3 marks] Which approach gives the better accuarcy on the test dataset?

**Options**

A) Approach 1

B) Approach 2

**Answer:**

* V1 - A
* V2 - A
* V3 - A
* V4 - A
* V5 - A


##Q13 [5 marks] What is the difference (absolute difference) in True Positives (TP) between these two approaches on test datasets?

**Answer**

V1: $68$

V2: $73$

V3: $1$

V4: $90$

V5: $32$

In [ ]:
y1_pred = DT_clf_1.predict(X_test)
y2_pred = DT_clf_2.predict(X_test)

cm_1 = confusion_matrix(y_test, y1_pred)
cm_2 = confusion_matrix(y_test, y2_pred)

print(abs(cm_1[1][1]- cm_2[1][1]))

[[ 780  243]
 [ 166 1311]]
[[ 775  248]
 [ 239 1238]]


## Q14 [5 marks] Find the sum of number of nodes of both trees (trees obtained using approach 1 and appoach 2).

V1: $386$

V2: $358$

V3: $350$

V4: $362$

V5: $346$

In [ ]:
DT_clf_1.tree_.node_count + DT_clf_2.tree_.node_count

358

# **(Common Instructions for Q15, and Q16)**
##Take a `RandomForestClassifier()` model with following hyperparameter values and tune it using GridsearchCV.

- Use n_estimators as [10,20,30]
- Use max_features as ['sqrt', 'log2']
- Use min_impurity_decrease as [0.001, 0.01, 0.1]
- Take cv value= 5 and random state = 64




In [ ]:

param_grid = {"n_estimators":[10, 20, 30], 'max_features': ['sqrt', 'log2'], 'min_impurity_decrease': [0.001, 0.01, 0.1]}

rf_gs = GridSearchCV(RandomForestClassifier(random_state = 64),
                     param_grid = param_grid, cv = 5)
rf_gs.fit(X_train,y_train)

rf_gs.best_params_

{'max_features': 'sqrt', 'min_impurity_decrease': 0.001, 'n_estimators': 30}

## Q15[5 marks] [NAT] Calculate the total number of misclassified samples by the best estimator for the test data.

Answer:

V1: 346   

V2: 393  

V3: 359  
  
V4: 332  
   
V5: 357

In [ ]:
cm = confusion_matrix(y_test, rf_gs.best_estimator_.predict(X_test))
cm[0][1] + cm[1][0]

393

## Q16 [5 marks] Find the value of min_impurity_decrease of the best estimator?

**Answer**

V1: 0.001

V2: 0.001

V3: 0.001

V4: 0.001

V5: 0.001

# **(Common Instructions for Q17,Q18)**
##Take an adaboost model with following hyperparameter values and tune it using GridsearchCV.

- Use n_estimators as [20,30, 50]
- Use learning_rate as [0.1, 0.5,1,2]
- Take cv value= 5 and random state = 64




In [ ]:
params = {'n_estimators':[20,30,50],
          'learning_rate':[0.1,0.5,1,2]}
ad_clf = AdaBoostClassifier(random_state = 64)
ad_gscv = GridSearchCV(ad_clf,
                       params,
                       cv=5)

ad_gscv.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=AdaBoostClassifier(random_state=64),
             param_grid={'learning_rate': [0.1, 0.5, 1, 2],
                         'n_estimators': [20, 30, 50]})

## Q17 [5 marks] Find the accuracy on the test dataset using the best estimator.

**Answer**

V1: $0.88$ Range: [0.84, 0.92]

V2: $0.87$ Range: [0.83, 0.91]

V3: $0.88$ Range: [0.84, 0.92]

V4: $0.88$ Range: [0.84, 0.92]

V5: $0.87$ Range: [0.83, 0.91]

In [ ]:
ad_gscv.best_estimator_.score(X_test, y_test)

0.8736

## Q18 [5 marks] How many class zero examples in the test dataset are correctly classified by the best estimator?

**Answer**

V1: 847

V2: 837

V3: 846

V4: 843

V5: 840

In [ ]:
confusion_matrix(y_test, ad_gscv.best_estimator_.predict(X_test))[0][0]

837


##Q19 [5 marks] Apply `GridSearchCV` using the support vector machine (SVM) classifier on the training dataset `X_train, y_train` and calculate the best value of `C` and `kernel` from the values below.

####use below hyperparameter values to tune the model
* `kernel':['linear', 'rbf'],
* 'C':[ 1, 10]`)
* Take cv =3 and random state to be 64

Which of the following options represent the best parameters?

options:

A. {'C': 1, 'kernel': 'rbf'}

B. {'C': 10, 'kernel': 'rbf'}

C. {'C':  1, 'kernel': 'linear'}

D. {'C': 10, 'kernel': 'linear'}


Answer:

C For all versions


In [ ]:
param_grid = {'kernel':('linear', 'rbf'), 'C':[1, 10]}

svc = SVC(random_state=64)

gs = GridSearchCV(svc, param_grid, cv = 3)

gs.fit(X_train, y_train)

gs.best_params_

{'C': 1, 'kernel': 'linear'}